libblkid/src/superblocks/bcachefs.c

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2021 Thomas Weißschuh <thomas@weissschuh.net>
+ *
+ * This file may be redistributed under the terms of the
+ * GNU Lesser General Public License.
+ *
+ * Based on code fragments from bcache-tools by Kent Overstreet:
+ * http://evilpiepirate.org/git/bcache-tools.git
+ * Also based on superblocks/bcache.c.
+ */
+
+#include "superblocks.h"
+
+#include <stddef.h>
+#include <stdio.h>
+
+#define BCACHEFS_SECTOR_SIZE   512
+#define BCACHEFS_SB_LABEL_SIZE 32
+
+struct bcachefs_super_block {
+	uint8_t		csum[16];
+	uint16_t	version;
+	uint16_t	version_min;
+	uint16_t	pad[2];
+	uint8_t		magic[16];
+	uint8_t		uuid[16];
+	uint8_t		user_uuid[16];
+	uint8_t		label[BCACHEFS_SB_LABEL_SIZE];
+	uint64_t	offset;
+	uint64_t	seq;
+	uint16_t	block_size;
+};
+
+/* magic string */
+#define BCACHEFS_SB_MAGIC     "\xc6\x85\x73\xf6\x4e\x1a\x45\xca\x82\x65\xf5\x7f\x48\xba\x6d\x81"
+/* magic string len */
+#define BCACHEFS_SB_MAGIC_LEN (sizeof(BCACHEFS_SB_MAGIC) - 1)
+/* super block offset */
+#define BCACHEFS_SB_OFF       0x1000
+/* supper block offset in kB */
+#define BCACHEFS_SB_KBOFF     (BCACHEFS_SB_OFF >> 10)
+/* magic string offset within super block */
+#define BCACHEFS_SB_MAGIC_OFF (offsetof(struct bcachefs_super_block, magic))
+
+static int probe_bcachefs(blkid_probe pr, const struct blkid_idmag *mag)
+{
+	struct bcachefs_super_block *bcs;
+
+	bcs = blkid_probe_get_sb(pr, mag, struct bcachefs_super_block);
+	if (!bcs)
+		return errno ? -errno : BLKID_PROBE_NONE;
+
+	blkid_probe_set_wiper(pr, 0, BCACHEFS_SB_OFF);
+	if (le64_to_cpu(bcs->offset) != BCACHEFS_SB_OFF / BCACHEFS_SECTOR_SIZE)
+		return BLKID_PROBE_NONE;
+
+	blkid_probe_set_uuid(pr, bcs->user_uuid);
+	blkid_probe_set_label(pr, bcs->label, sizeof(bcs->label));
+	blkid_probe_sprintf_version(pr, "%d", le16_to_cpu(bcs->version));
+	blkid_probe_set_block_size(pr, le16_to_cpu(bcs->block_size) * BCACHEFS_SECTOR_SIZE);
+	blkid_probe_set_uuid_as(pr, bcs->uuid, "INTERNAL_UUID");
```

Comment:
What is the purpose of the internal UUID and why we need it exported by libblkid?

Reply:
It is  "used for generating various magic numbers and identifying member devices, [never changes]".
I don't see an actual need to have it exported. I'll remove it.